addons/hr_timesheet/hr_timesheet.py

```diff
@@ -103,129 +77,118 @@ def on_change_unit_amount(self, cr, uid, id, prod_id, unit_amount, company_id, u
-        if context is None:
-            context = {}
-        emp_obj = self.pool.get('hr.employee')
-        emp_id = emp_obj.search(cr, uid, [('user_id', '=', context.get('user_id') or uid)], context=context)
-        ename = ''
-        if emp_id:
-            ename = emp_obj.browse(cr, uid, emp_id[0], context=context).name
-        if not vals.get('journal_id'):
-           raise UserError(_('No \'Analytic Journal\' is defined for employee %s \nDefine an employee for the selected user and assign an \'Analytic Journal\'!') % (ename,))
-        if not vals.get('account_id'):
-           raise UserError(_('No analytic account is defined on the project.\nPlease set one or we cannot automatically fill the timesheet.'))
-        return super(hr_analytic_timesheet, self).create(cr, uid, vals, context=context)
-
-    def on_change_user_id(self, cr, uid, ids, user_id):
-        if not user_id:
-            return {}
-        context = {'user_id': user_id}
-        return {'value': {
-            'product_id': self. _getEmployeeProduct(cr, uid, context),
-            'product_uom_id': self._getEmployeeUnit(cr, uid, context),
-            'general_account_id': self._getGeneralAccount(cr, uid, context),
-            'journal_id': self._getAnalyticJournal(cr, uid, context),
-        }}
-
-class account_analytic_account(osv.osv):
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Product")
+    @api.model
+    def _get_employee_unit(self, user_id=None):
+            emp = self.env['hr.employee'].search([('user_id', '=', self.user_id.id or user_id or self.env.uid)])
+            if emp and emp[0].product_id.uom_id.id:
+                return emp[0].product_id.uom_id.id
+            else:
+                raise exceptions.ValidationError("This user or employee is not associated to a valid Product Amount Type")    
+
+    @api.model
+    def _get_general_account(self, user_id=None):
+        emp = self.env['hr.employee'].search([('user_id', '=', self.user_id.id or user_id or self.env.uid)])
+        if emp and emp[0].product_id.categ_id.property_account_expense_categ.id:
+            return emp[0].product_id.categ_id.property_account_expense_categ.id
+        else:
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Product Financial Account")
 
+    @api.model
+    def _get_analytic_journal(self, user_id=None):
+        emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or user_id or self.env.uid)])
+        if emp and emp[0].journal_id.id:
+            return emp[0].journal_id.id
+        else:
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Journal ID")
+
+    is_timesheet = fields.Boolean()
+   
+    @api.constrains('user_id')
+    def check_user_id(self):
+        if self.is_timesheet:
+            emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or self.env.uid)])
```

Comment:
I don't think it is correct; here user_id should be correctly set on self. So I think you have to use directly self.user_id, and no fallback on self.env.uid. Moreover, add the api.one decorator, as it seems all constrains are api.one.